internal/daemon/api_logs.go

```diff
@@ -0,0 +1,318 @@
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+
+	// First write n buffered logs
+	if numLogs != 0 {
+		if len(itsByName) == 1 {
+			// Handle single-service case more efficiently
+			for _, it := range itsByName {
+				err = outputLogsSingle(w, it)
+			}
+		} else {
+			err = outputLogsMulti(w, itsByName, numLogs)
+		}
+		if err != nil {
+			fmt.Fprintf(w, "\nerror writing logs: %v", err)
+			return
+		}
+	}
+
+	// If requested, follow and output logs real time until request closed
+	if follow {
+		flushWriter(w)
+		followLogs(w, itsByName, req.Context().Done())
+	}
+}
+
+// Output a single iterator's logs without reading them all into memory.
+func outputLogsSingle(w io.Writer, it servicelog.Iterator) error {
+	reader := bufio.NewReaderSize(it, bufioReaderSize)
+	now := time.Now().UTC()
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+	err := iterateLogs(it, reader, nil, now, func(log logEntry) error {
+		return encoder.Encode(log)
+	})
+	return err
+}
+
+// Output last numLogs logs from multiple iterators (not particularly
+// efficient, as it writes them all to a slice first, then sorts by time).
+func outputLogsMulti(w io.Writer, itsByName map[string]servicelog.Iterator, numLogs int) error {
+	// Write all entries to slice.
+	var entries []logEntry
+	reader := bufio.NewReaderSize(nil, bufioReaderSize)
+	now := time.Now().UTC()
+	for _, it := range itsByName {
+		reader.Reset(it)
+		err := iterateLogs(it, reader, nil, now, func(log logEntry) error {
+			entries = append(entries, log)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Order by timestamp (or service name if timestamps equal).
+	sort.Slice(entries, func(i, j int) bool {
+		ei, ej := entries[i], entries[j]
+		if ei.Time.Equal(ej.Time) {
+			return ei.Service < ei.Service
+		}
+		return ei.Time.Before(ej.Time)
+	})
+
+	// Output (up to) the last numLogs logs from the sorted slice.
+	index := 0
+	if numLogs >= 0 && numLogs < len(entries) {
+		index = len(entries) - numLogs
+	}
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+	for _, log := range entries[index:] {
+		err := encoder.Encode(log)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Follow iterators and output logs as they're written.
+func followLogs(w io.Writer, itsByName map[string]servicelog.Iterator, done <-chan struct{}) {
+	// Start one goroutine per service to listen for and output new logs.
+	writeMutex := &sync.Mutex{}
+	var wg sync.WaitGroup
+	for service, it := range itsByName {
+		wg.Add(1)
+		go func(service string, it servicelog.Iterator) {
+			defer wg.Done()
+			reader := bufio.NewReaderSize(it, bufioReaderSize)
+			fw := &followWriter{w, writeMutex}
+			encoder := json.NewEncoder(fw)
+			encoder.SetEscapeHTML(false)
+			err := iterateLogs(it, reader, done, time.Time{}, func(log logEntry) error {
+				return encoder.Encode(log)
```

Comment:
This concurrency here feels a bit brittle. I'll comment more on why below, but as a suggested fix, it would be better if we managed to integrate this into the approach I described above, so that we're still sorting individual lines instead of entire buffers. Our ring buffer iterator is already a middle-man tracking positions between an in-memory buffer which is a moving target and the logic we have above. Maybe we can take that described algorithm and make it work even if the buffer reaches the current EOF.

Reply:
Concurrency: good call. I've updated this to send to a channel and write to the output from only a single goroutine. Safe now, and more Go-like in any case.